File: src/server/services/user.service.ts
import { eq } from "drizzle-orm";

import { db } from "~/server/db";
import { user } from "~/server/db/schema";

export type UserSummary = {
  id: string;
  email: string;
  name: string | null;
  image: string | null;
  emailVerified: boolean;
  createdAt: Date;
};

type UserRecord = typeof user.$inferSelect;

function mapUser(record: UserRecord): UserSummary {
  return {
    id: record.id,
    email: record.email,
    name: record.name,
    image: record.image,
    emailVerified: record.emailVerified,
    createdAt: record.createdAt,
  };
}

export async function getUserById(userId: string): Promise<UserSummary | null> {
  const [row] = await db
    .select({ user })
    .from(user)
    .where(eq(user.id, userId))
    .limit(1);

  if (!row) return null;

  return mapUser(row.user);
}

export async function getUserSafe(userId: string) {
  const userRecord = await getUserById(userId);
  if (!userRecord) {
    return null;
  }

  return userRecord;
}

export async function updateUserProfile(
  userId: string,
  data: { name?: string | null; image?: string | null },
): Promise<UserSummary | null> {
  const updatePayload: Partial<UserRecord> = {};

  if (data.name !== undefined) {
    updatePayload.name = data.name;
  }

  if (data.image !== undefined) {
    updatePayload.image = data.image;
  }

  if (Object.keys(updatePayload).length === 0) {
    return await getUserById(userId);
  }

  const [updated] = await db
    .update(user)
    .set(updatePayload)
    .where(eq(user.id, userId))
    .returning();

  if (!updated) {
    return null;
  }

  return mapUser(updated);
}

File: api/app/services/user_service.py
from typing import Any

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import User
from app.utils import iso


def map_user(record: User) -> dict[str, Any]:
    return {
        "id": record.id,
        "email": record.email,
        "name": record.name,
        "image": record.image,
        "emailVerified": record.email_verified,
        "createdAt": iso(record.created_at),
    }


def get_user_by_id(db: Session, user_id: str) -> dict[str, Any] | None:
    record = db.scalar(select(User).where(User.id == user_id).limit(1))
    if record is None:
        return None
    return map_user(record)


def get_user_by_email(db: Session, email: str) -> dict[str, Any] | None:
    record = db.scalar(select(User).where(User.email == email).limit(1))
    if record is None:
        return None
    return map_user(record)


def update_user_profile(
    db: Session, user_id: str, data: dict[str, Any]
) -> dict[str, Any] | None:
    record = db.scalar(select(User).where(User.id == user_id).limit(1))
    if record is None:
        return None

    changed = False
    if "name" in data and data["name"] is not None:
        record.name = data["name"]
        changed = True
    if "image" in data:
        record.image = data["image"]
        changed = True

    if changed:
        db.commit()
        db.refresh(record)

    return map_user(record)
